### PR TITLE
PWGHF: Make transition from McParticles_000 to McParticles_001

### PR DESCRIPTION
**Pull request by fgrosa**
@jgrosseo @vkucera @fcolamar please double check that I did not mess up anything. Thanks a lot!

### CODE REVIEW

File: Common/Core/RecoDecay.h
Comment (jgrosseo) [inline Common/Core/RecoDecay.h:717]: This could also become `auto particleI = arrDaughters[iProng].mcParticle();`

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx
Comment (jgrosseo) [inline PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:196]: `mcParticle()`

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:601]: How is `particleMother.globalIndex()` different from `particleMotherIdx`?

File: PWGHF/TableProducer/HFCandidateCreatorChic.cxx
Comment (jgrosseo) [inline PWGHF/TableProducer/HFCandidateCreatorChic.cxx:244]: @vkucera Shall we also change the syntax of this function? Converting internally now in a pair of numbers where `daughterIds()` is already a pair of numbers seems not needed...

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Done!

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx
Comment (fgrosa) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx]: Done! (everywhere)

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Indeed, fix it!

File: PWGHF/TableProducer/HFCandidateCreatorChic.cxx
Comment (jgrosseo) [reply PWGHF/TableProducer/HFCandidateCreatorChic.cxx:244]: Replace by just `has_daughters()`

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx
Comment (jgrosseo) [reply PWGHF/TableProducer/HFCandidateCreator3Prong.cxx]: `arrayDaughters[0].mcParticle().pdgCode()` did not work here?

File: Common/Core/RecoDecay.h
Comment (jgrosseo) [inline Common/Core/RecoDecay.h:734]: Replace by `has_daughters()`

Comment (jgrosseo) [reply Common/Core/RecoDecay.h:734]: Further below line 753 use `daughters().size()`

File: PWGHF/Tasks/HFMCValidation.cxx
Comment (jgrosseo) [inline PWGHF/Tasks/HFMCValidation.cxx:184]: Does `mother.daughters().front()` work here?

File: PWGHF/Tasks/taskD0parametrizedPID.cxx
Comment (jgrosseo) [inline PWGHF/Tasks/taskD0parametrizedPID.cxx:66]: Why not use the `using` in the argument list?

File: PWGHF/Tasks/taskD0parametrizedPID.cxx
Comment (jgrosseo) [inline PWGHF/Tasks/taskD0parametrizedPID.cxx:134]: This didn't work?

File: Common/Core/RecoDecay.h
Comment (vkucera) [reply Common/Core/RecoDecay.h:859]: `candidate.daughters().size()` would work the same, right?

**Comment by vkucera** [comment]
Actually, there are a few places in `RecoDecay` that would be benefit from replacement with `has_daughters()` and `.daughters().size()`.

File: PWGHF/Tasks/taskD0parametrizedPID.cxx
Comment (jgrosseo) [reply PWGHF/Tasks/taskD0parametrizedPID.cxx:134]: I mean `particle.daughter_as<McParticlesHf>().front().eta()`

File: PWGHF/Tasks/HFMCValidation.cxx
Comment (fgrosa) [reply PWGHF/Tasks/HFMCValidation.cxx:184]: This does not work, but I will update it with the method to retrieve the first daughter implemented in https://github.com/AliceO2Group/AliceO2/pull/8113

File: PWGHF/Tasks/taskD0parametrizedPID.cxx
Comment (fgrosa) [reply PWGHF/Tasks/taskD0parametrizedPID.cxx]: This does not work, but I will update it with the method to retrieve the first/last daughters implemented in https://github.com/AliceO2Group/AliceO2/pull/8113

**Comment by fgrosa** [comment]
> Actually, there are a few more places in `RecoDecay` that would benefit from replacement with `.has_daughters()` and `.daughters().size()`.

Should be done everywhere now!

File: PWGHF/Tasks/taskD0parametrizedPID.cxx
Comment (fgrosa) [reply PWGHF/Tasks/taskD0parametrizedPID.cxx:134]: The previous comment was not correct: as discussed in person I needed to use `begin()` instead of `front()` so it works now

File: PWGHF/Tasks/HFMCValidation.cxx
Comment (fgrosa) [reply PWGHF/Tasks/HFMCValidation.cxx]: The previous comment was not correct: as discussed in person I needed to use `begin()` instead of `front()` so it works now

**Comment by fgrosa** [comment]
@jgrosseo @vkucera the PR should be now updated using the methods implemented in https://github.com/AliceO2Group/AliceO2/pull/8113 and `rawIteratorAt()` instead of `iteratorAt()`

**Review by jgrosseo — APPROVED** [review]

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:599]: Revert. Variables starting with `particle` are meant to be particles everywhere else in the code.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:589]: Is this really the simplest way to get the first mother?

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:678]: Do not remove the debugging prints. They are useful whenever a new decay is being added.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:673]: This is the only place where the `particlesMC` table is used. Could it be removed by looping over `particle.daughters()` instead?

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:589]: This was suggested by @aalkin, I don't think there are easier ways to do it

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Reverted

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Right! I removed the argument and I replaced the loop over indices with a loop over particles directly

File: Common/Core/RecoDecay.h
Comment (aalkin) [reply Common/Core/RecoDecay.h:589]: All these is required because we need to call a dependent template inside a templated function, in `process()` one would not need the `template` keyword and type calculation, of course.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:671]: Looks like a line break was removed by accident.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:728]: Keep the debugging `Printf`.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:737]: Looks like a line break was removed by accident.

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:839]: Keep the debugging `Printf`.
```C++
//Printf("MC Gen: Rejected: no daughters");
```

File: Common/Core/RecoDecay.h
Comment (vkucera) [inline Common/Core/RecoDecay.h:843]: Keep the debugging `Printf`.

**Review by vkucera — APPROVED** [review]
The code looks fine. Just please fix the removed comments.

**Review by vkucera — CHANGES_REQUESTED** [review]
The code looks fine. Just please fix the removed comments.

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:849]: Done

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:844]: Done

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:740]: Fixed

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:730]: Done

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Fixed

File: Common/Core/RecoDecay.h
Comment (vkucera) [reply Common/Core/RecoDecay.h:740]: This has not been fixed yet. Instead the original lines were pasted below.

Comment (fgrosa) [reply Common/Core/RecoDecay.h]: RIght, I overlooked that, it should be now fixed!

**Review by vkucera — COMMENTED** [review]
Thanks a lot Fabrizio!
Please verify that the MC plots from the validation framework match the reference.

**Review by jgrosseo — DISMISSED** [review]

File: Common/Core/RecoDecay.h
Comment (fgrosa) [inline Common/Core/RecoDecay.h:676]: @vkucera when this piece of code is executed I get the following error:
`[ERROR] Exception caught: Unable to find column with label fIndexMcCollisions`
even if I checked and the column `fIndexMcCollisions` is present in the `O2mccollisionlabel` table in the `AO2D` file that I am analysing. 
I wrote this piece of code following what was suggested by @aalkin for 
`particleMother = particleMother.template mothers_first_as<typename std::decay_t<T>::parent_t>();` 
however I am not sure that it works also in this case. I also tried by replacing `typename std::decay_t<T>::parent_t` with simply `aod::McParticles` but I get the same error. @aalkin @jgrosseo do you have any suggestion? Thanks!

Comment (aalkin) [reply Common/Core/RecoDecay.h:676]: @fgrosa this particular exception is raised when constructing a table. When the slice index is unset it is supposed to return empty table. It may happen that this actually creates an empty table with no schema and thus the construction fails. For now you can protect this piece of code with check for `has_daughters()`.

Comment (fgrosa) [reply Common/Core/RecoDecay.h:676]: Hi @aalkin thanks! Unfortunately even if I put the protection:
```cpp
    if (particle.has_daughters()) {
      for (auto& dau : particle.template daughters_as<typename std::decay_t<T>::parent_t>()) {
        getDaughters(dau, list, arrPDGFinal, depthMax, stage);
      }
    }
```
it keeps crashing with the same error. Do you have any other idea by chance?

Comment (aalkin) [reply Common/Core/RecoDecay.h:676]: I will have to debug this, I assume the line `auto dauTable = particle.template daughters_as<typename std::decay_t<T>::parent_t>();` would crash as well, could you check?

Comment (vkucera) [reply Common/Core/RecoDecay.h:676]: Checking for `has_daughters()` has no effect here because this line is reached only if `has_daughters()` returned `true`.

Comment (fgrosa) [reply Common/Core/RecoDecay.h:676]: @aalkin indeed it crashes as well. @vkucera right because if `isFinal` is `true` and `has_daughters()` is false it goes through the first `if` statement, but the it returns anyway before this part of the code. 
I think that one possible solution would be to loop over indices in this case instead of getting the sliced table of daughters, but it would require to pass again the MC particle table as argument. However I have no better ideas at the moment. What do you think?

Comment (vkucera) [reply Common/Core/RecoDecay.h:676]: @aalkin Would this work?
```c++
  template <std::size_t N, typename T>
  static void getDaughters(const typename T::iterator& particle,
    ...
    for (auto& dau : particle.daughters_as<T>()) {
```

Comment (aalkin) [reply Common/Core/RecoDecay.h:676]: @vkucera I doubt this will be able to deduce type automatically from a function call, but it is better to check explicitly.

Comment (fgrosa) [reply Common/Core/RecoDecay.h:676]: @vkucera @aalkin I already tried this, but the compilation fails with the following error: `couldn't deduce template parameter 'T'`

**Comment by fgrosa** [comment]
Hi @jgrosseo @vkucera @aalkin with my last commit I reverted the code to the usage of `daughtersIds()` and `mothersIds()` in `RecoDecay.h` instead of `mothers_first_as<>` and `daughters_as<>` since I did not find a way to make it work otherwise. With this implementation I get the same result as compared to the reference efficiency in https://github.com/AliceO2Group/Run3Analysisvalidation/blob/master/codeHF/MC_d0_eff_ref.pdf so if you agree I would proceed with this now in the seek of moving forward (but indeed it would be good to understand why I did not manage to make it work otherwise). What do you think? Thanks a lot!

**Comment by jgrosseo** [comment]
I agree to move ahead and merge this. Could you subsequently, prepare a PR where you put `mothers_first_as` and `daughters_as` in a single location, such that @aalkin could follow up on it?

**Review by jgrosseo — DISMISSED** [review]

**Comment by jgrosseo** [comment]
Waiting for @vkucera approval of course...

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Fixed!

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: I added it for my debugging, but indeed it seems not necessary, removed

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h]: Fixed!

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:734]: `particleMother.daughtersIds().size()` is always 2 by definition because it's only first one and last one (suggested by @aalkin)

File: Common/Core/RecoDecay.h
Comment (fgrosa) [reply Common/Core/RecoDecay.h:854]: See reply above

**Comment by fgrosa** [comment]
> I agree to move ahead and merge this. Could you subsequently, prepare a PR where you put `mothers_first_as` and `daughters_as` in a single location, such that @aalkin could follow up on it?

Indeed this is a good idea @jgrosseo to keep working on it. I will open a new PR as soon as we merge this one

File: Common/Core/RecoDecay.h
Comment (vkucera) [reply Common/Core/RecoDecay.h:734]: OK, then the line below (735) should contain the same expression.

**Review by vkucera — APPROVED** [review]

File: PWGHF/Tasks/HFCorrelatorD0D0bar.cxx
Comment (vkucera) [inline PWGHF/Tasks/HFCorrelatorD0D0bar.cxx:454]: @fcolamar Can you please check these changes in your tasks and confirm that you are fine with them?

**Review by jgrosseo — APPROVED** [review]

**Pull request merged**
